cache/pmdk_cache.cc

```diff
@@ -0,0 +1,547 @@
+    th = GetTransientHandle(e, pack);
+    assert(e->InCache());
+    if (!e->HasRefs()){
+      LRU_Remove(e);
+    }
+    th->Ref();
+  }
+  return reinterpret_cast<Cache::Handle*>(th);
+}
+
+bool PMDKCacheShard::Ref(Cache::Handle* h) {
+  TransientHandle* e = reinterpret_cast<TransientHandle*>(h);
+  MutexLock l(&mutex_);
+  // To create another reference - entry must be already externally referenced
+  assert(e->HasRefs());
+  e->Ref();
+  return true;
+}
+
+void PMDKCacheShard::SetHighPriorityPoolRatio(double high_pri_pool_ratio) {
+  // TODO: call transient.
+}
+
+bool PMDKCacheShard::Release(Cache::Handle* handle, bool force_erase) {
+  if (handle == nullptr){
+    return false;
+  }
+  bool last_reference = false;
+  if (IsLRUHandle(handle)){
+    // TODO: release from transient tier
+
+    if (force_erase){
+      // TODO: erase in persistent tier
+      // grab the key in LRUHandle, find the persistent entry, and remove it.
+    }
+    // return false in this case, since
+    // there must be an entry in persistent tier.
+    last_reference = false;
+  } else {
+    bool* last_reference_p = &last_reference;
+    po::transaction::run(pop_, [&, handle, force_erase, last_reference_p] {
+      TransientHandle* e = reinterpret_cast<TransientHandle*>(handle);
+      {
+        MutexLock l(&mutex_);
+        *last_reference_p = e->Unref();
+        if (*last_reference_p && e->p_entry->InCache()) {
+          // The item is still in cache, and nobody else holds a reference to it
+          if (usage_ > persistent_capacity_ || force_erase){
+            // The LRU list must be empty since the cache is full
+            assert(lru_->next_lru == lru_ || force_erase);
+            // Take this opportunity and remove the item
+            persistent_hashtable_->Remove(e->p_entry);
+            e->p_entry->SetInCache(false);
+          } else {
+            // Put the item back on the LRU list, and don't free it
+            LRU_Insert(e->p_entry);
+            *last_reference_p = false;
+          }
+        }
+        if (*last_reference_p){
+          size_t total_charge = e->p_entry->persist_charge;
+          assert(usage_ >= total_charge);
+          usage_ -= total_charge;
+        }
+      }
+      if (*last_reference_p){
+        // TODO: double-check if this is able to free
+        // both persistent and transient metadata.
+        FreePEntry(e->p_entry);
+      }
+    });
+  }
+  return last_reference;
+}
+
+Status PMDKCacheShard::Insert(const Slice& key, uint32_t hash, void* value,
+                             size_t charge,
+                             void (*deleter)(const Slice& key, void* value),
+                             Cache::Handle** handle, Cache::Priority priority,
+                             const Slice (*unpack)(void* packed),
+                             void* (*pack)(const Slice& value)) {
+  
+  Status s;
+  Status* s_p = &s;
+
+  // TODO: insertion into transient tier, with a dummy deleter that
```

Comment:
If dummy deleter is used to insert to transient tier and insertion succeeds, which deleter will be used to free the `value` in the future?

Reply:
That's a very good question. After insertion, we want the transient tier to use the real deleter. I'll have to insert into persistent layer (or do the memcpy of `value`'s data) first, and probably return `Incomplete` immediately after failing to insert into persistent tier.

Reply:
Updated